### PR TITLE
Copter: Continue in LAND mode on failsafe event

### PR DESCRIPTION
**Pull request by Pedals2Paddles**
Adds new copter parameter `FS_LAND_CONTINUE` to enable/disable this feature.

If `FS_LAND_CONTINUE` is set for 1 (enabled) and the copter is in _Land Mode_ when radio, battery, or GCS failsafe activates, Copter will continue to land. If set for 0, Copter will carry out the failsafe action as usual.

The logic behind this is that if the vehicle is already in LAND mode, the operator has already positively decided to land at the given location. If a radio, GCS, or battery failsafe event happens _during that landing_, the vehicle should continue with the landing rather than carrying out the failsafe action.  Continuing to land gets the vehicle safely on the ground immediately. Whereas the full failsafe action, if set for RTL, at that point would just waste time and battery. The less time the vehicle spends in the air during a failsafe condition, the better. And if it's already trying to land, we might as well let it do that.

This was recently raised in https://github.com/ArduPilot/ardupilot/issues/6390.  

_Description updated 2017-11-22_

### CODE REVIEW

**Comment by peterbarker** [comment]
On Thu, 8 Jun 2017, Matt wrote:

> If the vehicle is already in LAND mode, the operator has already positively
> decided to land at the given location. If a radio or battery failsafe event

Not disagreeing with the patch (which I haven't reviewed, and I think I 
generally agree with not fail-safing out of LAND), but this isn't actually 
true.  The vehicle can enter LAND mode on an EKF failure, or if we breach 
the fence by >100m - and there are probably others.

**Comment by Pedals2Paddles** [comment]
This is true. I thought about most of those scenarios today, and I think I considered them all. But it's entirely possible I missed something that contraindicates this.  Here is my thoughts on those matters which I probably should have detailed in the PR to begin with:

* Operator enters land mode: Clearly the operator intended to land at the location anyway. This one is obvious. Doing exactly what the operator would expect it to do, land.

* EKF failsafe set to land: There would be no position information to carry out an RTL failsafe action anyway. So it would already be landing anyway, and continuing to land wouldn't change anything. So I believe it would continue doing exactly what the operator would expect it to do.

* 100m fence breach: If this is happening, trying to RTL or regain control has already failed.  Landing is basically it's last ditch effort to avoid flying off into lala land.  So continuing to land wouldn't actually change anything. It would continue doing exactly what the operator would expect it to do.

* Battery and radio failsafe actions sent to land mode, or when no GPS available: Mentioning this only because it exists. Land mode would be entered automatically by the failsafe. Obviously it's self-fulfilling and the end result is the same. Doing exactly what the operator would expect it to do.

* GCS failsafe with no GPS available: same as above. Doing exactly what the operator would expect it to do.

* Auto Mission landing: Technically it is in AUTO mode, not land mode. So this would not apply.

* RTL Mode landing: Technically it is in RTL mode, not land mode, so this would not apply.

**Comment by OXINARF** [comment]
Although I understand the desire for this behaviour, I disagree with it being the default, and specially I disagree with it not being optional.

> Operator enters land mode: Clearly the operator intended to land at the location anyway. This one is obvious. Doing exactly what the operator would expect it to do, land.

Not true, I've changed to Land mode before just to get the vehicle down without having to be controlling it and, when it was at the height I wanted, I changed back to Loiter. Even if the operator wanted to land, it doesn't mean he wants the vehicle to stay there if he can't control it anymore.

**Comment by geofrancis** [comment]
@OXINARF currently there is no way to land out side of rc range, even if your running out of battery, it will try to rtl as soon as it looses signal.

this is a good way of forcing a landing in emergancies like low battery, other aircraft in the area, etc.. 

it could be expanded so rather than just on or off it could be on if some failsafe conditions are met, ADSB, proximity, low battery im sure there could be some others im not thinking of.

**Comment by OXINARF** [comment]
@geofrancis 

> currently there is no way to land out side of rc range

Sure there is, GCS or setting up your failsafes to land.

> even if your running out of battery, it will try to rtl as soon as it looses signal.

If battery failsafe is active and is configured to land, a radio failsafe will not change mode and will continue to land.

> I used to have to hit the failsafe button on the radio module to reset it while in landing mode so that it wouldn't change on loss of rc but that only works with ppm since sbus has its signal lost flag.

I'm not sure what this means.

> keeping it in landing mode is a good way of forcing a landing in emergencies like low battery, other aircraft in the area, etc where rtl would normally kick in on loss of signal.

I'll quote myself:

> I understand the desire for this behaviour

That is completely different than forcing it on everyone. What you want is basically "I want battery failsafe to usually RTL, but if I'm already landing, land anyway" - that's a very specific behavior that I think we should try to accommodate, but certainly not impose on everyone.

**Comment by SergeyBokhantsev** [comment]
So maybe in scope of that fix you can add a new parameter? (option #2 https://github.com/ArduPilot/ardupilot/issues/6390 )
Just a thought, not sure what it is cost to add a new parameter.

**Comment by geofrancis** [comment]
no i want a way of landing outside RC range without having to mess with a laptop, RTL as low battery failsafe works 99% of the time without issue. but there are the rare occasions where it might not have the power to return, weather, failing battery etc, your saying that i should give up on RTL on low battery and set it to land on low battery failsafe. or risk it falling out the sky uncontrollably because it cant be landed without a RC signal.

**Comment by lvale** [comment]
two stage failsafe is a completely different issue, which is what should go in consideration. But 2 stage failsafe I can only see as its done by the Shenzen guys and it is on battery capacity. All other failsafes that can trigger either a RTL or Land would be unaffected by it.

**Comment by OXINARF** [comment]
>  your saying that i should give up on RTL on low battery and set it to land on low battery failsafe or risk it falling out the sky uncontrollably because it cant be landed without a RC signal.

You're saying that I should give up on RTL on RC lost signal if I'm in Land mode because you don't want *to mess with a laptop*. I gave you two options with the current code, you want me to have zero if this goes in...

> DJI machines have had 2 step battery failsafe since the first phantom, yet we have nothing equivalent on arducopter so thats why we need a way of forcing a landing.

So instead of contributing code to have two step battery failsafe, you think we should force everyone to Land if they lose RC signal and are in Land mode? I don't see how that makes sense.

**Comment by Pedals2Paddles** [comment]
I redo this PR with a new parameter, something like `FS_LAND_CONTINUE`, would that make everyone happy enough? :).  Set to 0 (default), it will behave as it does now, switching to RTL.  If enabled (set to 1), it will continue landing.

**Comment by OXINARF** [comment]
@Pedals2Paddles I'll find it acceptable - the parameter name should match what it does though, so if it only acts on RC failsafe calling it just FS_LAND_x doesn't look very good; if it acts on all failsafes, then it is a good name.

I should also say that the PR will always need to be approved by @rmackay9 as the Copter maintainer.

**Comment by Pedals2Paddles** [comment]
Good point. I'll come up with something meaningful.

**Comment by Kiwa21** [comment]
@Pedals2Paddles Any update on this ?
I think FS_LAND_CONTINUE would be the right option to go to. It doesn't change default behaviour for most users but it gives advanced users the opportunity to do more.

Yesterday, I had a different issue than landing in a remote aera : at the last stage of my flight I switch to land mode, 4m above ground my battery FS kicked in and copter changed into RTH which was in another location rather than continuing with landing. It wasn't very safe :)

**Comment by Pedals2Paddles** [comment]
Hoping to get back around to this later this week. I'm traveling for work and will have some down time with nothing else to do in the evenings.

**Comment by magicrub** [comment]
@Kiwa21 is referencing https://github.com/ArduPilot/ardupilot/pull/7033

**Comment by Pedals2Paddles** [comment]
Updated as per discussion.
* `FS_THR_LAND_CONT` now used to enable/disable continuing land mode when a radio failsafe happens
* Land mode will always continue on a battery failsafe event

**Comment by OXINARF** [comment]
> Land mode will always continue on a battery failsafe event

@Pedals2Paddles That's not what was discussed. What was discussed was keeping current behavior and have a parameter that allows changing that behavior.

File: ArduCopter/events.cpp [hunk 1]
```diff
@@ -16,8 +16,8 @@ void Copter::failsafe_radio_on_event()
     } else {
         if (control_mode == AUTO && g.failsafe_throttle == FS_THR_ENABLED_CONTINUE_MISSION) {
             // continue mission
-        } else if (control_mode == LAND && g.failsafe_battery_enabled == FS_BATT_LAND && failsafe.battery) {
```
Comment (OXINARF) [inline ArduCopter/events.cpp]: You can't remove those battery conditions.

**Comment by SergeyBokhantsev** [comment]
Just a note over here: there is a RC exists that cannot stop PPM feed but they just puts RC_3 channel to low value. So from ArduCopter prospective this isn't a radio failsafe but a manual switch to fly mode 6 (that is usually RTL). 
So for people with such HW setups there still no way to continue landing if RC lost.

**Comment by Pedals2Paddles** [comment]
> That's not what was discussed. What was discussed was keeping current behavior and have a parameter that allows changing that behavior.

The most recent discussion did not involve a parameter for the battery.  Since the operator still has control (unlike in radio failsafe), it doesn't really make sense change out of land mode. The operator wanted it in land mode. It's in land mode. And the operator still has control to do something else if they want.  I'm trying avoid parameter clutter but not making one for what seems logical anyway.

> Just a note over here: there is a RC exists that cannot stop PPM feed but they just puts RC_5 channel to low value. So from ArduCopter prospective this isn't a radio failsafe but a manual switch to fly mode 6 (that is usually RTL).
So for people with such HW setups there still no way to continue landing if RC lost.

Since that configuration is not properly using the failsafes, it is what it is.  That's not something that can be programmatically detected and acted upon.

File: ArduCopter/events.cpp (hunk 1 above)
Comment (Pedals2Paddles) [reply ArduCopter/events.cpp]: I read this over like 20 times and didn't notice what was going on here. Fixed.

File: libraries/AP_BattMonitor/AP_BattMonitor_Analog.cpp [hunk 2]
```diff
@@ -26,6 +26,11 @@ AP_BattMonitor_Analog::read()
 
     // get voltage
     _state.voltage = _volt_pin_analog_source->voltage_average() * _mon._volt_multiplier[_state.instance];
+    
+    // set watt capacity
+    if (_state.watts_capacity == 0 && _state.voltage_resting_estimate != 0 && _mon._pack_capacity != 0) {
+            _state.watts_capacity = _state.voltage_resting_estimate * (float)((uint32_t)_mon._pack_capacity / 1000);
```
Comment (SergeyBokhantsev) [inline libraries/AP_BattMonitor/AP_BattMonitor_Analog.cpp]: @Pedals2Paddles I think this is not the correct way to calculate Wh capacity.
Generally Wh value could be calculated as [capacity in Ah] * [average pack voltage]. Where average pack voltage is a constant, that is varies for different battery type (3.7 for 1S lipo).
But, even the approach above will not provide good accuracy. The only robust way is to have explicit setting watts_capacity. 
I mean user have to fly-out a battery then check his log for watts_used value. This will be the real value and this value need to set to watts_capacity.

**Comment by OXINARF** [comment]
@Pedals2Paddles 

> The most recent discussion did not involve a parameter for the battery

It did, otherwise we could have merged it... As I said there, the same objections I have for the RC failsafe, I have for any other failsafe.

> Since the operator still has control (unlike in radio failsafe), it doesn't really make sense change out of land mode.

You can't make that assumption. With current code it is possible that you don't have control and the vehicle is still flying (and with the path to full autonomy that's going to be more common).

>  The operator wanted it in land mode. It's in land mode.

And the operator wanted that if battery failsafe kicks in, the vehicle changes to RTL.

> I'm trying avoid parameter clutter but not making one for what seems logical anyway.

Why can't the same parameter be used for all? I would say that if an operator want Land to stay when RC and battery failsafe kicks in, then the operator wants that behaviour in all failsafes (even if it is just for RC and battery using the same parameter makes sense to me).

By the way, looks like a commit that doesn't belong here slipped into this PR.

**Comment by Pedals2Paddles** [comment]
I'll rework it for a single parameter.

**Comment by Pedals2Paddles** [comment]
Reconfigured for a single parameter for both failsafes. But I don't know travis is failing in SITL?

**Comment by khancyr** [comment]
It doesn't be related, it just failed to retreive ardurover version on SITL

**Comment by OXINARF** [comment]
@Pedals2Paddles @khancyr Either of you could have restarted the job. I've done it now.

**Comment by Pedals2Paddles** [comment]
How do you manually restart that?

**Comment by Pedals2Paddles** [comment]
Seems to be good now?

**Comment by OXINARF** [comment]
> How do you manually restart that?

@Pedals2Paddles Open Travis, login, press the "reload" button next to the job that failed. Looking for what the failure was is highly recommended before relaunching a job.

**Comment by Pedals2Paddles** [comment]
Rebased and made the last of the requested fixups.  This should be ready now @OXINARF  & @rmackay9

File: ArduCopter/Parameters.cpp [hunk 3]
```diff
@@ -1027,6 +1027,13 @@ const AP_Param::GroupInfo ParametersG2::var_info[] = {
     // @User: Standard
     AP_GROUPINFO("PILOT_SPEED_DN", 24, ParametersG2, pilot_speed_dn, 0),
 
+    // @Param: FS_LAND_CONTINUE
+    // @DisplayName: Failsafe land continue
+    // @Description: If FS_LAND_CONTINUE is set for RTL and the copter is in land mode when radio or battery failsafe activates, copter will continue to land. 0 = (default) Disabled (Always take failafe action).  1 = Enabled (Continue with land mode).
```
Comment (OXINARF) [inline ArduCopter/Parameters.cpp]: Sorry, but this is wrong now: *If **FS_LAND_CONTINUE** is set for RTL [...]*. That is this new parameter, it can't be set to RTL 😉

Since you have to modify it, please put _Copter_ and _Land_ with capital letter.

File: ArduCopter/events.cpp [hunk 4]
```diff
@@ -17,7 +17,9 @@ void Copter::failsafe_radio_on_event()
         if (control_mode == AUTO && g.failsafe_throttle == FS_THR_ENABLED_CONTINUE_MISSION) {
             // continue mission
         } else if (control_mode == LAND && g.failsafe_battery_enabled == FS_BATT_LAND && failsafe.battery) {
-            // continue landing
+            // continue landing because we're already in the battery failsafe landing phase
+        } else if (control_mode == LAND && g2.fs_land_continue) {
+            // continue landing because FS_THR_LAND_CONT is true
```
Comment (OXINARF) [inline ArduCopter/events.cpp]: FS_THR_LAND_CONT -> FS_LAND_CONTINUE

Or maybe better, just *[...] because parameter is true*

File: ArduCopter/events.cpp [hunk 5]
```diff
@@ -54,7 +56,9 @@ void Copter::failsafe_battery_event(void)
         if (should_disarm_on_failsafe()) {
             init_disarm_motors();
         } else {
-            if (g.failsafe_battery_enabled == FS_BATT_RTL || control_mode == AUTO) {
+            if (control_mode == LAND && g2.fs_land_continue) {
+                // continue landing because FS_THR_LAND_CONT is true
```
Comment (OXINARF) [inline ArduCopter/events.cpp]: Same.

**Comment by Pedals2Paddles** [comment]
Done.  While I was at it, I realized this should also apply to the GCS failsafe.  So I added the same logic that applies to the radio failsafe. Hope that's ok, it makes sense to me anyway.

**Comment by lvale** [comment]
Quick question.

As LAND is a flight mode and can be engaged at any time, what happens if a failsafe occurs during "normal" engagement of LAND ?

**Comment by geofrancis** [comment]
@lvale it continues to land, thats the point of this, if your already in landing mode it will continue if RC or Battery failsafe happens so it doesnt try and RTL.

**Comment by lvale** [comment]
@geofrancis If that is the case I don't think this behaviour should be implemented, because it makes the Land mode override a Failsafe situation.

**Comment by lvale** [comment]
@CraigElder Please schedule this for the next dev call.

**Comment by Pedals2Paddles** [comment]
> If that is the case I don't think this behaviour should be implemented, because it makes the Land mode override a Failsafe situation.

@lvale it is optional by parameter for that reason per discussion above. The PR description is old. I will update to be more descriptive of the current design.

**Comment by geofrancis** [comment]
@lvale I have had situations where there was not enough power to RTL and the only option was to land it out of radio range, but when i attempted this it would just RTL. there was nothing i could do other than to hover it as low as it would go and wait for the battery to fail and it to fall out the sky. 

this option give the pilot the ability to force a landing by engaging landing mode.

**Comment by lvale** [comment]
I'm sorry, but I can also imagine hundreds of situations where this would be good and as many where this would be bad....
It's a matter of consistency and principle.

Adding another parameter when we already have hundreds of parameters.....and this parameter would be a ignore failsafe on this flight mode....tomorrow we will be adding another set of parameters that would also ignore a failsafe if flying upside down, or during a flip disable failsafe, where do we stop.....

So a failsafe would only be valid  (IF/AND/OR)x(IF/AND/OR)x(IF/AND/OR) conditions/parameters, so it's better to just get rid of failsafes altogether..

**Comment by Pedals2Paddles** [comment]
I think that is an incredible hypothetical stretch without basis. There is a very pertinent and obvious use case for this. Ignoring it because someone might want to add a parameter to ignore failsafes while flying upside down, or whatever other ridiculous and useless case, is not meaningful.

This is why we have a dev team and PR process for peer review.  If someone submits an absurd patch to bypass failsafes for _no good reason_, there is a team of people here to point out its absurdity and deny the merge. They don't get an automatic free pass because this one made sense and got in.  The process here works, let it work.

If the issues is not wanting a new parameter, the solution becomes making the failsafe responses based on a bitmask. Which is way more complicated than a simple parameter.

File: ArduCopter/events.cpp (hunk 4 above)
Comment (OXINARF) [reply ArduCopter/events.cpp]: @Pedals2Paddles This wasn't fixed.

**Comment by Pedals2Paddles** [comment]
After the dev call discussion, it seemed that most people wanted to see a more robust failsafe options bitmask setup. That is a whole new thing, well over and above this minor patch.  Is this patch dead as a result?  Or is this going in?

**Comment by OXINARF** [comment]
@Pedals2Paddles I'm not sure what your doubt is. What was asked in the dev call was that instead of adding one parameter just for this option, it was done with a bitmask parameter. I understand if you aren't able to do that, but I think the result of the discussion was pretty clear.

**Comment by Pedals2Paddles** [comment]
It was clear that everyone wanted that. It was **not** clear if this was to be accepted in the meantime vs dead. If it was that clear, I wouldn't have asked.  Sorry to disturb you with such a silly question.... I'll close it.

**Comment by OXINARF** [comment]
@Pedals2Paddles I'm sorry if you felt offended, it wasn't my intention. I wrote *__I think__ the result was pretty clear*. It wasn't to be accepted as is, it was asked to be changed before being merged. There is no need to close this though so I'm re-opening it.

**Comment by Pedals2Paddles** [comment]
A complete rework of failsafe option handling is a much larger scope than simply continuing to land. Even if I were to take that on, it would be a whole new branch and PR with a much a larger scope.  This would be closed.

But, that is also way over my head to develop.  I can do basic stuff like this. But I can't completely rework the handling of failsafes and their options.  It's not that I don't want to, or don't like the idea. I want it, and think it's a great idea. It's just not something I'm able to do.  _So, that means someone else will have to do it.  Or nobody will do it, since it's not like everyone is sitting around bored with nothing else to do._

**Option One:** Merge this in the meantime, since who knows who or when someone else will completely rework the failsafe option handling.

**Option Two:** Don't merge this, and nobody with a use case for it it can have the feature.  Someone, somewhere sometime may totally rework the failsafe option handling. But until that happens, if that happens, we don't accept this, and presumably don't accept anything else similar?

Given the complete circumstances, option 1 makes more logical sense to me.  Unless someone is chomping at the bit to build the whole new thing, then obviously this can go away.

**Comment by OXINARF** [comment]
@Pedals2Paddles I know the discussion was long and several options were proposed, but in the end what was told as acceptable wasn't a complete rewrite of failsafes. It was simply having a FS_OPTIONS parameter where *if in Land mode, ignore settings and continue in Land* was one of the options - this is just so that when people have other edge cases they want supported we don't end up with 10 more parameters that 90% of people won't use.

The changes needed to this PR are quite minimal, it just takes a bit of time to grab it, edit, test, etc. I told you I was available to give you examples of where we already have option bitmask parameters in use.

**Comment by Pedals2Paddles** [comment]
There was talk of integrating the other existing failsafe options into it too.  Like RTL vs Land.  Continue in auto, etc.  That makes it way more than just a a different way to switch this land option.  It even got into discussing two-stage failsafes.  How much of that was wishing out loud vs "a decision" is not clear. 

Making this more than just turning a land option on and off is again way outside what I can tackle.  Making it just this land option, MIGHT be something I can tackle with a lot of help.

**Comment by geofrancis** [comment]
im a fan of option 1, there was a post of someone on the facebook group whos quad crashed after it engaged RTL and flew up into a tree while in landing mode so this is a needed request. i have had problems a couple of times, being unable to land remotely and having to let it fall out the sky, and having it try to climb to RTL altitude when in landing mode because it hit battery failsafe.

**Comment by OXINARF** [comment]
@Pedals2Paddles Yes, there was a discussion talking about all that - you've been in enough dev calls to know that discussions about what is ideal happen, but usually more practical solutions are needed and accepted. This was that case, we want to have a better failsafe system, but nobody is doing it now so it was accepted to do it just with a parameter.

> Making this more than just turning a land option on and off is again way outside what I can tackle.

Nobody asked you to do that.

> Making it just this land option, MIGHT be something I can tackle with a lot of help.

It's up to you. What I told you was that I can point you at existing examples so you can do it and get this merged faster. If you can't or won't do it, then someone will grab this PR and do it but there is no timeline on that.

**Comment by Pedals2Paddles** [comment]
Again, apologies for asking.  I'll see what I can do to redo it as described.

**Comment by Pedals2Paddles** [comment]
This: https://github.com/ArduPilot/ardupilot/blob/master/ArduCopter/events.cpp#L57

The battery failsafe setting is begin ignored and will always RTL if in auto mode.  If you have the battery failsafe disabled, or set to land, those settings are ignored and it will always RTL.  Is that desired?  It's not documented that I'm aware of. So I'd think people would expect it to do what the have it set for.  I'm on the fence whether this is a good thing or not.  I would personally want it to RTL, but that's what I have my battery failsafe set for too.  Someone with a use case for land only may not appreciate it doing the opposite?

**Comment by Pedals2Paddles** [comment]
I was on roll, so I kept going. This is now obsolete, in favor of the much large scope in https://github.com/ArduPilot/ardupilot/pull/7432.  Closing this one.
